qiskit/assertions/assertproduct.py

```diff
@@ -0,0 +1,133 @@
+
+
+        Args:
+            counts(dictionary): result.get_counts(experiment)
+
+        Returns:
+            tuple: tuple containing:
+
+                chisq(float): the chi-square value
+                pval(float): the p-value
+                passed(bool): if the test passed
+        """
+        q0len = len(self._qubit0)
+        q1len = len(self._qubit1)
+        cont_table = np.empty((2 ** q0len, 2 ** q1len))
+        cont_table.fill(1)
+        for (key, value) in counts.items():
+            key_rev = key[::-1]
+            q0index = int(key_rev[:q0len], 2)
+            q1index = int(key_rev[q0len:], 2)
+            cont_table[q0index][q1index] = value
+
+        chisq, pval, _, _ = chi2_contingency(cont_table)
+        passed = bool(pval >= self._pcrit)
+        return (chisq, pval, passed)
+
+
+def get_breakpoint_product(self, qubit0, cbit0, qubit1, cbit1, pcrit=0.05):
+    """
+    Creates a breakpoint, which is a renamed deep copy of the QuantumCircuit, and creates and
+    appends an AssertProduct instruction to its end.  It tests whether qubit0 and qubit1, measured
+    to cbit0 and cbit1, have any entanglement between them.  If the statistical test passes, the
+    assertion passes; if the test fails, the assertion fails.
+
+    Args:
+        qubit0(QuantumRegister or list or tuple): quantum register
+        cbit0(ClassicalRegister or list or tuple): classical register
+        qubit1(QuantumRegister or list or tuple): quantum register
+        cbit1(ClassicalRegister or list or tuple): classical register
+        pcrit(float): critical p-value for the hypothesis test
+
+    Returns:
+        QuantumCircuit: copy of quantum circuit at the assert point
+    """
+    clone = self.copy(Asserts.breakpoint_name())
+    assertion = AssertProduct(qubit0, cbit0, qubit1, cbit1, pcrit, False)
+    clone.append(assertion, [assertion._qubit], [assertion._cbit])
+    return clone
+
+
+QuantumCircuit.get_breakpoint_product = get_breakpoint_product
+
+
+def get_breakpoint_not_product(self, qubit0, cbit0, qubit1, cbit1, pcrit=0.05):
+    """
+    Creates a breakpoint, which is a renamed deep copy of the QuantumCircuit, and creates and
+    appends an AssertProduct instruction to its end.  It tests whether qubit0 and qubit1, measured
+    to cbit0 and cbit1, have any entanglement between them.  If the statistical test passes, the
+    assertion fails; if the test fails, the assertion passes.
+
+    Args:
+        qubit0(QuantumRegister or list or tuple): quantum register
+        cbit0(ClassicalRegister or list or tuple): classical register
+        qubit1(QuantumRegister or list or tuple): quantum register
+        cbit1(ClassicalRegister or list or tuple): classical register
+        pcrit(float): critical p-value for the hypothesis test
+
+    Returns:
+        QuantumCircuit: copy of quantum circuit at the assert point
+    """
+    clone = self.copy(Asserts.breakpoint_name())
```

Comment:
if `self` is a QuantumCircuit, why not change it in-place? `get_breakpoint_not_product` could be called `append_breakpoint_not_product`.

Reply:
A breakpoint is a deepcopy of the QuantumCircuit, with an assertion instruction tacked onto the end.  The reason for copying the QuantumCircuit is to allow multiple breakpoints and the original QuantumCircuit to be executed.  Since the assertion is a measure instruction, appending (instead of copying and appending) would alter the behavior of the remainder of the QuantumCircuit.